File: abc089/abc089_d.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
  int h,w,d;
  cin >> h >> w >> d;

  map<int,pair<int,int>> mp;
  for (int i = 0; i < h; i++)
  {
    for (int j = 0; j < w; j++)
    {
      int a;
      cin >> a;
      mp[a] = make_pair(i,j);
    }    
  }

  vector<int> dis(h*w+5);
  for (int i = 1; i < h*w+5; i++)
  {
    if(i>d){
      dis[i]=dis[i-d]+abs(mp[i].first-mp[i-d].first) + abs(mp[i].second-mp[i-d].second);
    }
  }  

  int q;
  cin >> q;
  for (int i = 0; i < q; i++)
  {
    int l,r;
    cin >> l >> r;
    int ans = dis[r]-dis[l];
    cout << ans << endl;
  }
}

File: abc089/abc089_d.rb
h,w,d = gets.split.map(&:to_i)
ha={}
h.times do |i|
  ar=gets.split.map(&:to_i)
  w.times do |j|
    ha[ar[j]]=[i,j]
  end
end

dis=Array.new(h*w+1,0)
(1..h*w).each do |i|
  if i>d
    dis[i]=dis[i-d]+(ha[i].first-ha[i-d].first).abs+(ha[i].last-ha[i-d].last).abs
  end
end

gets.to_i.times do
  l,r=gets.split.map(&:to_i)
  puts dis[r]-dis[l]
end